graylog2-server/src/main/java/org/graylog2/inputs/codecs/JsonPathCodec.java

```diff
@@ -105,6 +134,50 @@ protected String buildShortMessage(Map<String, Object> fields) {
         return shortMessage.toString();
     }
 
+    public Map<String, Object> useFlattener(String json) throws JsonTypeException {
+        Map<String, Object> map = new HashMap<>();
+        try {
+            flatten("", OBJECT_MAPPER.readTree(json), map);
```

Comment:
Instead of parsing a mutable object (`map`), the `flatten` method you are calling here could just return an immutable map.